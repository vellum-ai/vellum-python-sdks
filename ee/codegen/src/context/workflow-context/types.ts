export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
