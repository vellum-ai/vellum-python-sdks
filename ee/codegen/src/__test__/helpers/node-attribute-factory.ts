import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
