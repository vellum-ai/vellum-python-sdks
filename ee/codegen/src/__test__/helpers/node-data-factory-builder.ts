import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
