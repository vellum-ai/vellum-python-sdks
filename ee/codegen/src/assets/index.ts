// This directory contains generated assets that are copied during postinstall
// and included in the npm package build. The node-definitions.json file is
// generated from the Python SDK and copied here for use in code generation.
