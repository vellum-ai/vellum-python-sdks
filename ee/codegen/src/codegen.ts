import {
  ErrorLogFile,
  InitFile,
  Inputs,
  NodeInput,
  VellumValue,
  VellumVariable,
  Workflow,
} from "./generators";
import { BasePersistedFile } from "./generators/base-persisted-file";
import { State } from "./generators/state";
import { WorkflowSandboxFile } from "./generators/workflow-sandbox-file";

export function vellumVariable(args: VellumVariable.Args): VellumVariable {
  return new VellumVariable(args);
}

export function vellumValue(args: VellumValue.Args): VellumValue {
  return new VellumValue(args);
}

export function nodeInput(args: NodeInput.Args): NodeInput {
  return new NodeInput(args);
}

export function inputs(args: Inputs.Args): Inputs {
  return new Inputs(args);
}

export function state(args: State.Args): State {
  return new State(args);
}

export function workflow(args: Workflow.Args): Workflow {
  return new Workflow(args);
}

export function initFile(args: InitFile.Args): InitFile {
  return new InitFile(args);
}

export function errorLogFile(args: BasePersistedFile.Args): ErrorLogFile {
  return new ErrorLogFile(args);
}

export function workflowSandboxFile(
  args: WorkflowSandboxFile.Args
): WorkflowSandboxFile {
  return new WorkflowSandboxFile(args);
}
