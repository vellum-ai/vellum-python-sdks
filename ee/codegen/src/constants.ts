/* Module paths */
export const VELLUM_CLIENT_MODULE_PATH = ["vellum"] as const;
export const VELLUM_WORKFLOW_NODE_BASE_TYPES_PATH = [
  "vellum",
  "workflows",
  "nodes",
  "displayable",
  "bases",
  "types",
] as const;
export const VELLUM_WORKFLOW_GRAPH_MODULE_PATH = [
  "vellum",
  "workflows",
  "graph",
] as const;
export const VELLUM_WORKFLOW_NODES_MODULE_PATH = [
  "vellum",
  "workflows",
  "nodes",
] as const;
export const VELLUM_WORKFLOW_CONSTANTS_PATH = [
  "vellum",
  "workflows",
  "references",
  "constant",
] as const;
export const VELLUM_WORKFLOWS_DISPLAY_BASE_PATH = [
  "vellum_ee",
  "workflows",
  "display",
  "base",
] as const;
export const VELLUM_WORKFLOW_EDITOR_TYPES_PATH = [
  "vellum_ee",
  "workflows",
  "display",
  "editor",
] as const;
/* Class names */
export const OUTPUTS_CLASS_NAME = "Outputs";
export const PORTS_CLASS_NAME = "Ports";
export const DEFAULT_PORT_NAME = "default";

/* File names */
export const INIT_FILE_NAME = "__init__.py";

export const GENERATED_WORKFLOW_MODULE_NAME = "workflow";

export const GENERATED_INPUTS_MODULE_NAME = "inputs";

export const GENERATED_NODES_MODULE_NAME = "nodes";

export const GENERATED_DISPLAY_MODULE_NAME = "display";
export const GENERATED_NODES_PATH = [GENERATED_NODES_MODULE_NAME] as const;

export const GENERATED_DISPLAY_NODE_MODULE_PATH = [
  GENERATED_DISPLAY_MODULE_NAME,
  GENERATED_NODES_MODULE_NAME,
] as const;

export const GENERATED_NESTED_NODE_MODULE_NAME = "node";
