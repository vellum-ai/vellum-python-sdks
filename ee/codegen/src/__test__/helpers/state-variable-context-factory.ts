import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
