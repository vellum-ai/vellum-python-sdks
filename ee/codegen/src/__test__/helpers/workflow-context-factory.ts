import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
