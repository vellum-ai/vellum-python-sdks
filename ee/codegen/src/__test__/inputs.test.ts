import { Writer } from "@fern-api/python-ast/core/Writer";
import { Vellum } from "vellum-ai";
import { VellumVariable } from "vellum-ai/api";
import { VellumVariableType } from "vellum-ai/api/types";

import { workflowContextFactory } from "./helpers";
import { inputVariableContextFactory } from "./helpers/input-variable-context-factory";

import * as codegen from "src/codegen";
import { WorkflowContext } from "src/context";

describe("Inputs", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();
  });

  describe("write", () => {
    it("should generate correct code when Inputs has no variables", async () => {
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code when Inputs has variables", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "input1", key: "input1", type: "STRING", required: true },
        { id: "input2", key: "input2", type: "NUMBER", required: true },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code when Inputs has variables with empty keys", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "input1", key: "", type: "STRING", required: true },
        { id: "input2", key: "", type: "STRING", required: true },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code when Inputs has variables with duplicate keys", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "input1", key: "key-1", type: "STRING", required: true },
        { id: "input2", key: "key-1", type: "STRING", required: true },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code when Inputs has a custom name", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "input1", key: "input1", type: "STRING", required: true },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({
        workflowContext,
        name: "CustomInputs",
      });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should not generate any code when Inputs is empty", async () => {
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toBe("");
    });

    it("should generate correct code for complex input variables", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "1", key: "query", type: "STRING", required: true },
        { id: "2", key: "max_runtime", type: "NUMBER", required: true },
        {
          id: "3",
          key: "previous_chat_history",
          type: "CHAT_HISTORY",
          required: true,
        },
        {
          id: "4",
          key: "prior_results",
          type: "SEARCH_RESULTS",
          required: true,
        },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({
        workflowContext,
      });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should convert input variable names into valid python attributes", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "1", key: "My Input", type: "STRING", required: true },
        { id: "2", key: "$My*Input", type: "NUMBER", required: true },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({
        workflowContext,
      });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it.each([
      {
        type: VellumVariableType.String,
        key: "string_input",
        default: {
          type: "STRING",
          value: "Example String",
        } as Vellum.StringVellumValue,
      },
      {
        type: VellumVariableType.Number,
        key: "number_input",
        default: { type: "NUMBER", value: 123 } as Vellum.NumberVellumValue,
      },
      {
        type: VellumVariableType.Json,
        key: "json_input",
        default: {
          type: "JSON",
          value: { key: "value" },
        } as Vellum.JsonVellumValue,
      },
      {
        type: VellumVariableType.Image,
        key: "image_input",
        value: {
          type: "IMAGE",
          default: {
            src: "IMAGE",
          } as Vellum.VellumImage,
        } as Vellum.ImageVellumValue,
      },
      {
        type: VellumVariableType.Audio,
        key: "audio_input",
        value: {
          type: "AUDIO",
          default: {
            src: "AUDIO",
          } as Vellum.VellumAudio,
        } as Vellum.AudioVellumValue,
      },
      {
        type: VellumVariableType.FunctionCall,
        key: "function_call_input",
        default: {
          type: "FUNCTION_CALL",
          value: {
            arguments: { arg1: "Hello World" },
            name: "function_call",
          } as Vellum.FunctionCall,
        } as Vellum.FunctionCallVellumValue,
      },
      {
        type: VellumVariableType.Error,
        key: "error_input",
        default: {
          type: "ERROR",
          value: {
            message: "Some 500 error",
            code: Vellum.VellumErrorCodeEnum.InternalServerError,
          } as Vellum.VellumError,
        } as Vellum.ErrorVellumValue,
      },
      {
        type: VellumVariableType.Array,
        key: "array_input",
        default: {
          type: "ARRAY",
          value: [
            {
              type: "STRING",
              value: "Example String",
            } as Vellum.StringVellumValue,
            { type: "NUMBER", value: 123 } as Vellum.NumberVellumValue,
          ],
        } as Vellum.ArrayVellumValue,
      },
      {
        type: VellumVariableType.ChatHistory,
        key: "chat_history_input",
        default: {
          type: "CHAT_HISTORY",
          value: [
            {
              text: "foo bar",
              role: Vellum.ChatMessageRole.User,
            } as Vellum.ChatMessage,
          ],
        } as Vellum.ChatHistoryVellumValue,
      },
      {
        type: VellumVariableType.SearchResults,
        key: "search_results_input",
        default: {
          type: "SEARCH_RESULTS",
          value: [
            {
              text: "Hello, World!",
              score: 1.0,
              keywords: ["foo", "bar"],
              document: {
                label: "Example Document",
              } as Vellum.SearchResultDocument,
            } as Vellum.SearchResult,
          ],
        } as Vellum.SearchResultsVellumValue,
      },
    ])(
      "should generate correct code when default is a $type",
      async ({ type, key, default: defaultValue }) => {
        const extensions: Vellum.VellumVariableExtensions = {
          color: "red",
        };
        const inputVariables: VellumVariable[] = [
          {
            id: "some-id",
            key,
            type,
            required: true,
            default: defaultValue,
            extensions: extensions,
          },
        ];

        inputVariables.forEach((inputVariableData) => {
          workflowContext.addInputVariableContext(
            inputVariableContextFactory({
              inputVariableData,
              workflowContext,
            })
          );
        });

        const inputs = codegen.inputs({ workflowContext });

        inputs.write(writer);
        expect(await writer.toStringFormatted()).toMatchSnapshot();
      }
    );

    it("should generate correct Optional types when required is false", async () => {
      const inputVariables: VellumVariable[] = [
        {
          id: "1",
          key: "foo",
          type: "STRING",
          default: { type: "STRING", value: "Example String" },
          required: false,
        },
        {
          id: "2",
          key: "bar",
          type: "NUMBER",
          default: { type: "NUMBER", value: 123 },
          required: false,
        },
      ];

      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });

      const inputs = codegen.inputs({ workflowContext });
      inputs.write(writer);

      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
