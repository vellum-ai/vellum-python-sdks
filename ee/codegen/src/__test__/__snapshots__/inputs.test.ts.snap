// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inputs > write > should convert input variable names into valid python attributes 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import Union


class Inputs(BaseInputs):
    my_input: str
    my_input_1: Union[float, int]
"
`;

exports[`Inputs > write > should generate correct Optional types when required is false 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import Optional, Union


class Inputs(BaseInputs):
    foo: Optional[str] = "Example String"
    bar: Optional[Union[float, int]] = 123
"
`;

exports[`Inputs > write > should generate correct code for complex input variables 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import Union, List
from vellum import ChatMessage, SearchResult


class Inputs(BaseInputs):
    query: str
    max_runtime: Union[float, int]
    previous_chat_history: List[ChatMessage]
    prior_results: List[SearchResult]
"
`;

exports[`Inputs > write > should generate correct code when Inputs has a custom name 1`] = `
"from vellum.workflows.inputs import BaseInputs


class CustomInputs(BaseInputs):
    input1: str
"
`;

exports[`Inputs > write > should generate correct code when Inputs has no variables 1`] = `""`;

exports[`Inputs > write > should generate correct code when Inputs has variables 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import Union


class Inputs(BaseInputs):
    input1: str
    input2: Union[float, int]
"
`;

exports[`Inputs > write > should generate correct code when Inputs has variables with duplicate keys 1`] = `
"from vellum.workflows.inputs import BaseInputs


class Inputs(BaseInputs):
    key_1: str
    key_1_1: str
"
`;

exports[`Inputs > write > should generate correct code when Inputs has variables with empty keys 1`] = `
"from vellum.workflows.inputs import BaseInputs


class Inputs(BaseInputs):
    input_: str
    input_1: str
"
`;

exports[`Inputs > write > should generate correct code when default is a 'ARRAY' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import List
from vellum import VellumValue


class Inputs(BaseInputs):
    array_input: List[VellumValue] = [
        "Example String",
        123,
    ]
"
`;

exports[`Inputs > write > should generate correct code when default is a 'AUDIO' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from vellum import VellumAudio


class Inputs(BaseInputs):
    audio_input: VellumAudio
"
`;

exports[`Inputs > write > should generate correct code when default is a 'CHAT_HISTORY' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import List
from vellum import ChatMessage


class Inputs(BaseInputs):
    chat_history_input: List[ChatMessage] = [
        ChatMessage(role="USER", text="foo bar"),
    ]
"
`;

exports[`Inputs > write > should generate correct code when default is a 'ERROR' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from vellum import VellumError


class Inputs(BaseInputs):
    error_input: VellumError = VellumError(
        message="Some 500 error", code="INTERNAL_SERVER_ERROR"
    )
"
`;

exports[`Inputs > write > should generate correct code when default is a 'FUNCTION_CALL' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from vellum import FunctionCall


class Inputs(BaseInputs):
    function_call_input: FunctionCall = FunctionCall(
        arguments={
            "arg1": "Hello World",
        },
        name="function_call",
    )
"
`;

exports[`Inputs > write > should generate correct code when default is a 'IMAGE' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from vellum import VellumImage


class Inputs(BaseInputs):
    image_input: VellumImage
"
`;

exports[`Inputs > write > should generate correct code when default is a 'JSON' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import Any


class Inputs(BaseInputs):
    json_input: Any = {
        "key": "value",
    }
"
`;

exports[`Inputs > write > should generate correct code when default is a 'NUMBER' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import Union


class Inputs(BaseInputs):
    number_input: Union[float, int] = 123
"
`;

exports[`Inputs > write > should generate correct code when default is a 'SEARCH_RESULTS' 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import List
from vellum import SearchResult, Document


class Inputs(BaseInputs):
    search_results_input: List[SearchResult] = [
        SearchResult(
            text="Hello, World!",
            score=1,
            keywords=["foo", "bar"],
            document=Document(id="", label="Example Document"),
        ),
    ]
"
`;

exports[`Inputs > write > should generate correct code when default is a 'STRING' 1`] = `
"from vellum.workflows.inputs import BaseInputs


class Inputs(BaseInputs):
    string_input: str = "Example String"
"
`;
